src/borrow_tracker/tree_borrows/mod.rs

```diff
@@ -285,42 +313,65 @@ trait EvalContextPrivExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
         let span = this.machine.current_span();
         let alloc_extra = this.get_alloc_extra(alloc_id)?;
-        let range = alloc_range(base_offset, ptr_size);
         let mut tree_borrows = alloc_extra.borrow_tracker_tb().borrow_mut();
 
-        // All reborrows incur a (possibly zero-sized) read access to the parent
-        if new_perm.initial_read {
+        let mut perform_parent_access = |range: AllocRange| -> InterpResult<'tcx> {
             tree_borrows.perform_access(
                 orig_tag,
                 Some((range, AccessKind::Read, diagnostics::AccessCause::Reborrow)),
                 this.machine.borrow_tracker.as_ref().unwrap(),
                 alloc_id,
                 this.machine.current_span(),
-            )?;
-        }
+            )
+        };
+
+        // Some reborrows incur a read access to the parent
+        let NewPermission { freeze_access, nonfreeze_access, .. } = new_perm;
+        this.visit_freeze_sensitive(place, ptr_size, |mut range, frozen| {
+            range.start += base_offset;
+            let access = if frozen { freeze_access } else { nonfreeze_access };
+            if access {
+                perform_parent_access(range)?
+            }
+            interp_ok(())
+        })?;
+
         // Record the parent-child pair in the tree.
         tree_borrows.new_child(
+            this,
+            place,
+            ptr_size,
+            base_offset,
             orig_tag,
             new_tag,
-            new_perm.initial_state,
-            range,
+            new_perm,
             span,
-            new_perm.protector.is_some(),
         )?;
         drop(tree_borrows);
 
-        // Also inform the data race model (but only if any bytes are actually affected).
-        if range.size.bytes() > 0 && new_perm.initial_read {
+        let perform_initial_access = |range: AllocRange| -> InterpResult<'tcx> {
             if let Some(data_race) = alloc_extra.data_race.as_vclocks_ref() {
                 data_race.read(
                     alloc_id,
                     range,
                     NaReadType::Retag,
                     Some(place.layout.ty),
                     &this.machine,
-                )?;
+                )?
             }
-        }
+            interp_ok(())
+        };
+
+        // Also inform the data race model (but only if any bytes are actually affected).
+        let NewPermission { freeze_access, nonfreeze_access, .. } = new_perm;
+        this.visit_freeze_sensitive(place, ptr_size, |mut range, frozen| {
```

Comment:
What is going on here? We should definitely not call `visit_freeze_sensitive` twice.

Reply:
I tried to put the `perform_parent_access` part into the
`visit_freeze_sensitive` call in `new_child`, so we can remove this
`visit_freeze_sensitive` call, but this causes some tests to fail.

Reply:
So what is the reasons for those tests to fail?